### PR TITLE
[Feature] Add RoF2 Bazaar Support

### PR DESCRIPTION
**Pull request by neckkola**
# Description

This PR will re-write much of the bazaar functionality with the intent to be phase 1 of RoF2 support.  Phase 2 will be to enable the barter/buyer system.  I am separating the two to allow for easier testing/review.  I have marked this draft, as I believe it is ready for folks to kick the tires and comment upon.  More changes will occur as I test further and respond to any feedback, suggestions, bugs, etc.  I trust this is a good approach.

Specifically, this change;
- rewrites much of the encode/decode functions to be base on action keys instead of packet length
- rewrites much, though not all, of the pre-existing merchant/bazaar functionality, again to be based on action instead of packet length.
- the change to action allows for backwards compatibility with Ti, UF, and RoF2
- Enables /bazaar functionality for RoF2
- Enable co-existence with UF/RoF2 traders including the 80/200 limits
- supports tradeable augmented items
- supports RoF2 clients to purchase via parcel delivery outside of bazaar at a delivery cost.  I set this to 20% as a default.  Not sure what live is atm.
- creates several rules to adjust parcel delivery cost(implemented) and voucher delivery (not yet implemented)

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

# Testing
I have tested via RoF2 and UF.  Ti compatibility is not yet complete.  I will tackle this next.

Clients tested: 
RoF2 and UF

# Checklist

- [X] I have tested my changes
- [X] I have performed a self-review of my code. Ensuring variables, functions and methods are named in a human-readable way, comments are added only where naming of variables, functions and methods can't give enough context.
- [TODO] I have made corresponding changes to the documentation (if applicable, if not delete this line)
- [X] I own the changes of my code and take responsibility for the potential issues that occur
- [X] If my changes make database schema changes, I have tested the changes on a local database (attach image). Updated version.h CURRENT_BINARY_DATABASE_VERSION to the new version. (Delete this if not applicable)

# Comments

- Ti integration not complete.
- The client, at least with the old implementation of the bazaar, and with these updates, has a limitation where duplicate item ids within a trader will have the same sale price.  For example, if you have two stacks of item 8517 (Class 6 Arrows) they will be the same price.  I will investigate this more, though this is how it previously functioned, so I have not spent much time at the moment.
- Augmented items are supported.  Due to the client issue above, if you have two items, one augmented and one not, they will have the same price.
- RoF2 direct delivery via vouchers is not enabled.  Are expedient delivery vouchers enabled within alt currency?  I can look into this if folks would like this functionality.

Any suggests, or feedback are welcome, including any testing that people could do.

Thanks, and enjoy.
Neckkola

### CODE REVIEW

**Comment by neckkola** [comment]
> Can we completely disable parcel delivery and force users to visit the bazaar stalls?

I can add that. No problem

File: common/database/database_update_manifest.cpp [hunk 1]
```diff
@@ -5635,6 +5635,33 @@ ALTER TABLE `npc_spells_entries` ADD `content_flags_disabled` varchar(100) NULL;
...
+	ADD COLUMN `aug_slot_5` INT(10) UNSIGNED NOT NULL DEFAULT '0' AFTER `aug_slot_4`,
+	ADD COLUMN `aug_slot_6` INT(10) UNSIGNED NOT NULL DEFAULT '0' AFTER `aug_slot_5`,
+	CHANGE COLUMN `serialnumber` `item_sn` INT(10) UNSIGNED NOT NULL DEFAULT '0' AFTER `aug_slot_6`,
+	CHANGE COLUMN `charges` `item_charges` INT(11) NOT NULL DEFAULT '0' AFTER `item_sn`,
+	ADD COLUMN `char_entity_id` INT(11) UNSIGNED NOT NULL DEFAULT 0 AFTER `slot_id`,
+	ADD COLUMN `char_zone_id` INT(11) UNSIGNED NOT NULL DEFAULT 0 AFTER `char_entity_id`,
+	ADD COLUMN `active_transaction` TINYINT(3) UNSIGNED NOT NULL DEFAULT 0 AFTER `char_zone_id`,
+	DROP PRIMARY KEY,
+	ADD PRIMARY KEY (`id`),
+	ADD INDEX `Index 2` (`char_id`, `slot_id`);
```
Comment (Akkadius) [inline common/database/database_update_manifest.cpp:5662]: Can we give this a meaningful name like `charid_slotid`

**Comment by neckkola** [comment]
> > Can we completely disable parcel delivery and force users to visit the bazaar stalls?
> 
> I can add that. No problem

@fryguy503 Added a new rule.  Default is to allow delivery so you will have to flip to false.

File: common/database/database_update_manifest.cpp (hunk 1 above)
Comment (neckkola) [reply common/database/database_update_manifest.cpp:5662]: Updated.

**Comment by neckkola** [comment]
Good afternoon.  I have worked through the remaining bazaar search items.  Everything should now be working in search except for prestige items as I understand that they are not implemented.  Others should work; item name, local/all/specific traders, race, class, stats, type, level, cost and number of returned items.

I have also now added Titanium support for both /trader and /bazaar.

It a previous version has been running in PEQ for several days, no reported issues that I am aware of, other than the limited search functionality.  Should be ready to go.

I will need to do some cleanup as well, as there are some unused functions, duplicate enums.  Will look after this for sure, then add some documentation, videos.

Neckkola

**Comment by Valorith** [comment]
> Good afternoon.  I have worked through the remaining bazaar search items.  Everything should now be working in search except for prestige items as I understand that they are not implemented.  Others should work; item name, local/all/specific traders, race, class, stats, type, level, cost and number of returned items.
> 
> 
> 
> I have also now added Titanium support for both /trader and /bazaar.
> 
> 
> 
> A previous version has been running in PEQ for several days, no reported issues that I am aware of, other than the limited search functionality.  Should be ready to go.
> 
> 
> 
> I will need to do some cleanup as well, as there are some unused functions, duplicate enums.  Will look after this for sure, then add some documentation, videos.
> 
> 
> 
> Neckkola

You're a legend!

**Comment by fryguy503** [comment]
Can I add a request?

Currently to use barter, you must use the new bazaar. Before I had hardcoded a location in the old bazaar stalls to use barter.

Now with using the old bazaar, we need a door down for players to stand on. Can we code this in as an option for barter. So a different door to stand on will trigger barter allowance?

In era there was two old bazaars one for traders and one for barters, that did not last long before the new and arguably worse bazaar was added.

**Comment by Akkadius** [comment]
Been testing on PEQ for a good week, been solid

**Review by Akkadius — APPROVED** [review]

**Pull request merged**